Mutagen optimization

yeegor opened the pull request:
Bring the project up using Mutagen.
TODO: in closest time make docker-compose files to be able to bring the project up without Mutagen.

yeegor committed:
Fixing certificate revoked on Mac

yeegor committed:
Use Mutagen to bring up core config

yeegor committed:
Wait for necessary files to be present when
building PWA theme.

yeegor committed:
Remove whitespaces

yeegor committed:
Mac optimizations
- Add volume core-frontend-data for core frontend container
- Implement mutagen configuration for both core and frontend dc configs
- FIx node container not having wget
- Keep maildev configuration in local setup only

yeegor committed:
Optimize frontend deploy script acquiring process
Make them be mounted into frontend container on deploy
instead of copied on build

yeegor committed:
Add mutagen config without frontend container
Optimize sync making it two-way-resolved
Propagate www-data as default file owner

yeegor committed:
Move core-frontend-data volume from main compose
to core compose

yeegor committed:
Fix permission issue in app and nginx containers

yeegor committed:
Rename mutagen compose files to keep originals

yeegor committed:
Add modified original compose files back

yeegor committed:
Move core frontend volume declaration
to corresponding compose file

yeegor committed:
Fix user being set as file owner on alpha

yeegor commented:
For now I am unable to install backend module using this setup. Investigating.

yeegor committed:
Refactor sync sessions
- remove nginx session by introducing named volume
- extend mutagen permission config with default file mode

alfredsgenkins commented on build/php/Dockerfile:
```diff
@@ -90,6 +90,7 @@ RUN apt-get -qq update &&\
     xz-utils \
     libssl-dev \
     libreadline-dev \
+    iputils-ping \
```
I think this is not used anymore.

yeegor replied:
Removed.

alfredsgenkins commented on deploy/shared/conf/local-ssl/ca.conf:
```diff
@@ -17,8 +17,8 @@ serial = serial
-default_crl_days = 3650
-default_days = 3650
+default_crl_days = 825
+default_days = 825
```
Please add comments why we did this

yeegor replied:
I did that to fix certificate issue on Mac, which did no help. Reverting.

alfredsgenkins commented on deploy/start-demo.sh:
```diff
@@ -65,14 +65,17 @@ function install_additional_composer_modules {
-  # Theme build
-  if [ $? -eq 0 ]; then
+
+  if ping -c 1 frontend &> /dev/null
```
We still have this? I thought we removed it!

alfredsgenkins replied:
Because we have additional .sh file for Frontend.

eli-l replied:
Demo has nothing common with frontend container, why there are changes here?

yeegor replied:
returned it back.

alfredsgenkins commented on deploy/start.sh:
```diff
@@ -14,7 +14,7 @@ fi
 ### Debug options end
 
-### Custom commands for assets compilation
+### Customdc commands for assets compilation
```
Typo?

alfredsgenkins commented:
Please try adding `container_name: my-web-container` for all setvices docker-compose.mutagen.yml or in docker-compose.local.yml to resolve naming issue. Note: this prevents from scaling.

alfredsgenkins commented:
Please link a doc for this, otherwise Ilja won't review this.

alfredsgenkins commented:
Try specifying user in docker-compose like this: https://docs.docker.com/compose/compose-file/#userns_mode#domainname-hostname-ipc-mac_address-privileged-read_only-shm_size-stdin_open-tty-user-working_dir

eli-l commented on build/frontend/start.sh:
```diff
@@ -1,4 +1,14 @@
 #!/bin/bash
+
+cd app/design/frontend/Scandiweb/pwa/
```
You've changed the mountpoint / workdir, why?

yeegor replied:
It is done in order not to sync huge app-data volume with frontend container, who needs like 2% of the information contained in it, it speeds the process up.

eli-l replied:
What is "like 2%"?

eli-l replied:
Seems I won't get an expected answer, but it is up to nginx config file to control what is served and what is not, you must not decide on which files to copy/sync and which not, as this will lead to misbehavior.
First of all - it is not up to sync mechanism to decide which files are available to nginx and which to fpm.
Second - this will lead to different behavior with mutagen and without it. It is not the way this should be implemented. Make the behavior consistent!

yeegor replied:
Returned mounting all the app-data volume into the frontend container.

yeegor committed:
Revert not working Mac certificate fix

eli-l commented on deploy/start-with-frontend.sh:
```diff
@@ -0,0 +1,315 @@
+#!/usr/bin/env bash
```
Should I use this shell script to start the stack with frontend container or...? File name is completely unclear here.

yeegor replied:
Yes, you should use this shell script to start the stack with frontend container. That is why it is called "start-with-frontend". I am open to your proposals regarding the naming of this file.

eli-l replied:
"start-when-frontend-serving.sh" - would clearly provide use case of this file, isn't it?

yeegor replied:
It does as well, renamed.

yeegor committed:
Typo and ping removal

yeegor committed:
Remove ping from installed packages

eli-l commented on deploy/start.sh:
```diff
@@ -57,17 +57,16 @@ export bash_colors
-    cd $BASEPATH
-  fi
+
+  echo "${blue}${bold}Building PWA theme${normal}"
```
Why the check was removed?

yeegor replied:
Because now there are two start scripts: 'start' which builds the theme inside of app container and 'start with frontend' which builds the theme in the frontend container. The deploy is twice as fast with this configuration because theme is able to be compiled asynchronously in frontend container while app is getting up if it is needed.

eli-l replied:
I don't care about the speed if functionality is broken. Instead of not building if setup:upgrade returned non-zero code - you built a script that does not care abt Magento status.
start.sh is a default entrypoint, why do you make some changes there refering frontend container etc. FE containers got own files.

eli-l replied:
Get check back, this is the default file used in most of environments.

yeegor replied:
Returned the check.

eli-l commented on docker-compose.yml:
```diff
@@ -53,7 +54,7 @@ services:
       - ./deploy/shared/conf/nginx/magento.conf:/etc/nginx/conf.d/default.conf
       - ./deploy/shared/conf/nginx/cache-router.conf:/etc/nginx/conf.d/cache-router.conf
-      - app-data:/var/www/public
+      - nginx-data:/var/www/public/pub
```
What's the reason of introducing new volume? Is that mutagen limitation?

yeegor replied:
It is not mutagen-related, but apparently it is faster to sync many small named volumes with their containers than one huge app-data volume with many containers.

yeegor replied:
It is not the only volume introduced, see core-frontend-data.
I also wanted to create a frontend-data volume to mount it into the frontend container, but the issues occur because we then would need to mount vendor and app/design/frontend/scandiweb/pwa folders separately and vendor doesn't want to be mounted separately for some reason :(

eli-l replied:
> It is not mutagen-related, but apparently it is faster to sync many small named volumes with their containers than one huge app-data volume with many containers.

Ok, means that you prefer "faster sync" to developer, who is adding additional file and want it to be served directly from nginx, to spend TONS of time to find out it is not mounted to nginx? This approach is wrong.

yeegor replied:
Current solution considers mounting app-data into nginx container. This way we will not have one more docker-compose file. Is there any difference in mounting app-data volume into nginx or creating a bind mount? Additional sync sessions will be slower than direct named volume mount, should I create them between nginx and local fs instead of mounting app-data volume into nginx?

yeegor committed:
Add mutagen-powered-app container name
to resolve container naming issue in yml files

yeegor commented:
Known issue: Currently I am experiencing problems with configuring Xdebug on my Mac using this configuration. 
Investigating.

yeegor commented:
> Try specifying user in docker-compose like this: https://docs.docker.com/compose/compose-file/#userns_mode#domainname-hostname-ipc-mac_address-privileged-read_only-shm_size-stdin_open-tty-user-working_dir

When I refer to user 'user' by name, the container does not start.
When I refer to user 'user' by UID 1000, I get the following error:
`mutagen-powered-app | Starting with UID : 0
mutagen-powered-app | useradd: Permission denied.
mutagen-powered-app | useradd: cannot lock /etc/passwd; try again later.
mutagen-powered-app | error: failed switching to "user": unable to find user user: no matching entries in passwd file`

eli-l commented on build/frontend/Dockerfile:
```diff
@@ -8,9 +8,4 @@ ARG BASEPATH=/var/www/public
 # Set working directory so any relative configuration or scripts wont fail
 WORKDIR $BASEPATH
 
```
Why these scripts were removed? These files contains deployment logic, not application version/dependencies.

yeegor replied:
Returned a single script to build frontend container

eli-l commented on build/frontend/start-core.sh:
```diff
@@ -1,4 +1,12 @@
 #!/bin/bash
+
+echo "Waiting for initial files to be synced"
```
I'd be more up to enforce re-building when dependencies are changed, rather than to use such a hacky solution. It is needed not because Docker is not-to-good, but because we are moving away from it's best practices to far.

yeegor replied:
Do you propose to do install dependencies during the frontend container build and remove `npm ci` from this script at all, leaving it with a single purpose of launching pm2-runtime? Then we can get rid of these deployment scripts of frontend container and do that with docker-compose file `command: ['npm', 'run', 'pm2-watch']` directive?

yeegor replied:
`npm ci` was moved to frontend container build script.

eli-l commented on build/frontend/start.sh:
```diff
@@ -1,4 +1,14 @@
 #!/bin/bash
```
Since you've been touching start.sh and start-core.sh - this is the place we could use some "optimization" and remove the "core" file, utilizing environment variables and docker-compose instead, as all "core" configs are meant for core developers only, and we could share the knowledge quickly.
The "unification" could be done with envvar, that would indicate which config file to use. This would be good improvement.

yeegor replied:
Implemented.

eli-l commented on build/rendertron/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM node:10-slim
```
Why base image has been changed? How it is related to mutagen?

yeegor replied:
It is changed in the 2.x-stable as well to resolve issues with wget not being available inside the container. See the following commit: https://github.com/scandipwa/scandipwa-base/commit/f672efcafeb431dcfd75cec25e1e773e8d4c16ef

eli-l commented on deploy/start-demo.sh:
```diff
@@ -65,20 +65,18 @@ function install_additional_composer_modules {
-    npm run build
-    cd $BASEPATH
-  fi
+
```
As already mentioned - why this check has been removed? There is not sense to build a theme if Migration failed. Why you're changing the deployment logic?

yeegor replied:
I have returned this check into start-demo script. Script that starts with frontend container has no need in this check because theme is being built asynchronously, independent of app container.

eli-l commented on deploy/start-demo.sh:
```diff
@@ -286,6 +284,12 @@ function exit_catch {
 
 ### Deploy pipe start
 
+echo "${blue}${bold}Waiting for Mutagen to sync initial filese${normal}"
```
Composer.json/lock is always in a container, as these are added during the build.
Does mutagen add additional layer that is empty on top of containers FS, similar to local mount? Even so - this behavior is related only to local.

eli-l replied:
For mutagen - use another start scripts, hopefully we'll unify it one day.

eli-l commented on deploy/start.sh:
```diff
@@ -264,6 +263,13 @@ function exit_catch {
 
 ### Deploy pipe start
 
+echo "${blue}${bold}Waiting for Mutagen to sync initial filese${normal}"
```
This is the default file - nothing related to local, mutagen or other side-cases must be added here.
create deploy/mutagen directory, put there everything related to Mutagen sync, anyway you have mutagen dedicated docker-compose files, controlling start script is extremely easy there.

eli-l commented on docker-compose.core.yml:
```diff
@@ -14,9 +14,15 @@ services:
     volumes:
       - ./src/localmodules/base-theme:/var/www/public
-    command: ["npm", "run", "dev-server-core"]
+      - ./build/frontend/start-core.sh:/start-core.sh
```
Build file into container, do not mount it from the host.

eli-l reviewed (changes requested):
Global comment:
- pay attention and rework workdirs, don't rework/redefine them unless absolutely necessary for Mutagen (never?)
- source code must be shared between nginx and app containers, no matter how you run it (no mounting, docker native, mutagen), it is done for a reason.
- revisit all the changes (all existing files that have been edited by you), and unless this is absolutely good for ALL environments and use cases - move these changes to a separate file (even it is 300 lines long).
- remove changes to the existing services, that are not related to your task

yeegor committed:
Refactor frontend container compose files
Move dependency installation to build script
Replace deploy script with single command
Remove core-frontend-data volume, mount app-data directly
Rename start-with-frontend script

yeegor committed:
Mount all app-data into nginx

yeegor committed:
Return magento se:up success check to deploy

yeegor committed:
Update process file names to the relevant ones

yeegor committed:
Return dependency installation to FE deploy

yeegor committed:
Revert demo deploy script changes

yeegor committed:
Use separate deploy script for Mutagen
Remove waiting for composer.json files in main script

yeegor committed:
Use start-with-mutagen script to wait for fs

yeegor committed:
Map Mutagen start scripts into containers

yeegor committed:
Build scripts into containers instead of map

yeegor committed:
Mount app-data into  frontend container

yeegor committed:
Remove mounting start script from host

yeegor committed:
Move waiting logic to separate script for FE cont

yeegor committed:
Apply proper permissions to waiting script

yeegor committed:
Implement script to wait for fs

yeegor committed:
Merge branch 'mutagen-optimization' of github.com:yeegor/scandipwa-base into mutagen-optimization
